### PR TITLE
Texture caching enhancements for Web Viewer

### PR DESCRIPTION
**Pull request by kwokcb**
## Texture Caching Enhancements

- Add in texture cache reuse using ThreeJS cache. 
- Handle http/blob/data encodings properly. ( This will allow support for interactive image assignment
to existing shaders. )

### Implementation Changes

- Change the texture loader logic to check in the ThreeJS cache first to avoid duplicate loads.
- Extend URI support to accept http/blob/data encodings for input images. Blobs are not cached since they reference transient data used by the active web page.

### Example
- Uses `http://` address of a MaterialX logo as image input. 
- Viewer result shown below after using drag-drop to load it in. The document is also available from the documents menu.

```xml
<?xml version="1.0"?>
<materialx version="1.39" colorspace="lin_rec709">
  <surface_unlit name="Default_Unlit" type="surfaceshader">
    <input name="emission" type="float" value="1" />
    <input name="emission_color" type="color3" nodename="tiledimage_color4" />
    <input name="transmission" type="float" value="0" />
    <input name="transmission_color" type="color3" value="1,1,1" />
    <input name="opacity" type="float" value="1" />
  </surface_unlit>
  <surfacematerial name="Default" type="material">
    <input name="surfaceshader" type="surfaceshader" nodename="Default_Unlit" />
    <input name="backsurfaceshader" type="surfaceshader" value="" />
    <input name="displacementshader" type="displacementshader" value="" />
  </surfacematerial>
  <tiledimage name="tiledimage_color4" type="color3">
    <input name="file" type="filename" value="https://materialx.org/images/MaterialXLogoSmallA.png" colorspace="none" />
    <input name="default" type="color3" value="0,0,0" />
    <input name="texcoord" type="vector2" value="0,0" />
    <input name="uvtiling" type="vector2" value="1,1" />
    <input name="uvoffset" type="vector2" value="0,0" />
    <input name="realworldimagesize" type="vector2" value="1,1" unittype="distance" />
    <input name="realworldtilesize" type="vector2" value="1,1" unittype="distance" />
    <input name="filtertype" type="string" value="linear" />
    <input name="framerange" type="string" value="" />
    <input name="frameoffset" type="integer" value="0" />
    <input name="frameendaction" type="string" value="constant" />
  </tiledimage>
</materialx>
```

image

### Example
- Using embedded data. In this case a 1 pixel pink square. The document is available from the documents menu.
```xml 
<?xml version="1.0"?>
<materialx version="1.39" colorspace="lin_rec709">
  <surface_unlit name="image_shader" type="surfaceshader">
    <input name="emission" type="float" value="1" />
    <input name="emission_color" type="color3" nodename="colorspaced_image" />
    <input name="transmission" type="float" value="0" />
    <input name="transmission_color" type="color3" value="1,1,1" />
    <input name="opacity" type="float" value="1" />
  </surface_unlit>
  <surfacematerial name="image_material" type="material">
    <input name="surfaceshader" type="surfaceshader" nodename="image_shader" />
    <input name="backsurfaceshader" type="surfaceshader" value="" />
    <input name="displacementshader" type="displacementshader" value="" />
  </surfacematerial>
  <image name="colorspaced_image" type="color3">
    <input name="file" type="filename" value="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAIAAACQd1PeAAAADElEQVQI12P4z/AfAAQAAf/zKSWvAAAAAElFTkSuQmCC" colorspace="srgb_texture" />
    <input name="layer" type="string" value="" />
    <input name="default" type="color3" value="0,0,0" />
    <input name="texcoord" type="vector2" value="0,0" />
    <input name="uaddressmode" type="string" value="periodic" />
    <input name="vaddressmode" type="string" value="periodic" />
    <input name="filtertype" type="string" value="linear" />
    <input name="framerange" type="string" value="" />
    <input name="frameoffset" type="integer" value="0" />
    <input name="frameendaction" type="string" value="constant" />
  </image>
</materialx>
```
image

### Example
- This is an example with a blob. It won't load the image since it was taken from another session but shows the format:
```xml
<?xml version="1.0"?>
<materialx version="1.39" colorspace="lin_rec709">
  <surface_unlit name="image_shader" type="surfaceshader">
    <input name="emission" type="float" value="1" />
    <input name="emission_color" type="color3" nodename="colorspaced_image" />
    <input name="transmission" type="float" value="0" />
    <input name="transmission_color" type="color3" value="1,1,1" />
    <input name="opacity" type="float" value="1" />
  </surface_unlit>
  <surfacematerial name="image_material" type="material">
    <input name="surfaceshader" type="surfaceshader" nodename="image_shader" />
    <input name="backsurfaceshader" type="surfaceshader" value="" />
    <input name="displacementshader" type="displacementshader" value="" />
  </surfacematerial>
  <image name="colorspaced_image" type="color3">
    <input name="file" type="filename" value="blob:http://localhost:8000/74cf5fec-38cb-4335-a1f1-21769e7e9c35" colorspace="srgb_texture" />
    <input name="layer" type="string" value="" />
    <input name="default" type="color3" value="0,0,0" />
    <input name="texcoord" type="vector2" value="0,0" />
    <input name="uaddressmode" type="string" value="periodic" />
    <input name="vaddressmode" type="string" value="periodic" />
    <input name="filtertype" type="string" value="linear" />
    <input name="framerange" type="string" value="" />
    <input name="frameoffset" type="integer" value="0" />
    <input name="frameendaction" type="string" value="constant" />
  </image>
</materialx>
```
This is how it looks in the session it was used in. 
image

### CODE REVIEW

File: javascript/MaterialXView/source/helper.js [hunk 1]
```diff
@@ -117,18 +117,65 @@ function toThreeUniform(type, value, name, uniforms, textureLoader, searchPath,
...
-                const texture = textureLoader.load(fullPath);
-                // Set address & filtering mode
-                if (texture)
-                    setTextureParameters(texture, name, uniforms, flipY);
-                outValue = texture;
+                // Cache / reuse texture to avoid reload overhead.
+                // Note: that data blobs and embedded data textures are not cached as they are transient data.
+                let checkCache = true;
+                let texturePath = searchPath + IMAGE_PATH_SEPARATOR + value;
+                if (value.startsWith('blob:'))
```
Comment (kwokcb) [inline javascript/MaterialXView/source/helper.js:124]: Blobs, embedded data and http references are no relative to a search path. This fixes it so users can use these texture URI specifications in their files.

File: javascript/MaterialXView/source/helper.js [hunk 2]
```diff
@@ -117,18 +117,65 @@ function toThreeUniform(type, value, name, uniforms, textureLoader, searchPath,
...
+                    texturePath = value;
+                    console.log('Load HTTP URL:', texturePath);
+                }
+                else if (value.startsWith('data:'))
+                {
+                    texturePath = value;
+                    checkCache = false;
+                    console.log('Load data URL:', texturePath);
+                }
+                const cachedTexture = checkCache && THREE.Cache.get(texturePath);
```
Comment (kwokcb) [inline javascript/MaterialXView/source/helper.js:141]: Do proper cache checking.

File: javascript/MaterialXView/source/index.js [hunk 3]
```diff
@@ -179,8 +186,19 @@ function animate()
         viewer.getScene().setUpdateTransforms();
     }
 
-    renderer.render(viewer.getScene().getScene(), viewer.getScene().getCamera());
-    viewer.getScene().updateTransforms();
+    // Only re-render when an update request was made
```
Comment (kwokcb) [inline javascript/MaterialXView/source/index.js:189]: Re-render gated by requirement to refresh. All the `schedulUpdate()` calls triggers the requirment.

File: javascript/MaterialXView/source/viewer.js [hunk 4]
```diff
@@ -622,7 +627,12 @@ export class Material
 
         // Load material
         if (mtlxMaterial)
-            await mx.readFromXmlString(doc, mtlxMaterial, searchPath);
+            try {
```
Comment (kwokcb) [inline javascript/MaterialXView/source/viewer.js:630]: Small fix. Would "crash" sometimes on invalid files.

**Review by jstone-lucasfilm — CHANGES_REQUESTED** [review]
I really like all of the small fixes and improvements in this change, but I'm less convinced by the caching scheme, which runs counter to the rendering approach I see in other performance-focused web applications.  Additionally, the caching scheme has the side effect of making `viewer` an input to `toThreeUniform`, which doesn't seem ideal.

What would you think about preserving all of the small fixes and improvements in this change, but omitting the caching scheme for now, and we can consider this idea separately in the future?

**Comment by kwokcb** [comment]
Sounds like a good separation. Maybe let's table looking at why "render" has so much overhead. It could be draw time, it could be uniform bind time or state setup time or something else. OpenPBR is the slowest shader but std surface is also good for testing.

**Comment by kwokcb** [comment]
@jstone-lucasfilm, This is purely a texture caching update PR now.

**Review by jstone-lucasfilm — COMMENTED** [review]
This looks good to me, @kwokcb!  Is there a generated web site where we can validate these changes in the context of the viewer?

**Review by jstone-lucasfilm — APPROVED** [review]

**Comment by kwokcb** [comment]
Hi @jstone-lucasfilm, I added a "data" example that can be used in the viewer. The "blob" case I can't really make it show unless the viewer page was loading the data. (I just showed it in my own editor which uses the same URI support but has UI to support setting local images -- basically what Felix logged an issue for that `MaterialXViewer` could support).

**Comment by jstone-lucasfilm** [comment]
@kwokcb Oh, I wasn't suggesting that you should add anything more to the set of fixes that I reviewed.  I was just curious as to whether you had a generated web site where I could validate that the latest code worked well.

**Comment by kwokcb** [comment]
Ah. My mistake. I reverted the last push. I tried publishing the viewer page from branch but didn't have any luck.

**Comment by jstone-lucasfilm** [comment]
Thanks @kwokcb, and I'll give this a try in my own fork of MaterialX.

**Pull request merged**